src/systems/camera-system.js

```diff
@@ -86,33 +132,79 @@ export class CameraSystem {
 
   inspect(o) {
     if (this.mode !== CAMERA_MODE_INSPECT) {
-      this.preInspectMode = this.mode;
+      this.snapshot.mode = this.mode;
     }
     this.mode = CAMERA_MODE_INSPECT;
     this.inspected = o;
+
+    (getBatch(o, this.batchManagerSystem) || o).traverse(enableInspectLayer);
+
+    const scene = AFRAME.scenes[0];
+    const vrMode = scene.is("vr-mode");
+    const camera = vrMode ? scene.renderer.vr.getCamera(scene.camera) : scene.camera;
+    this.snapshot.mask = camera.layers.mask;
+    camera.layers.set(CAMERA_LAYER_INSPECT);
+    if (vrMode) {
+      this.snapshot.mask0 = camera.cameras[0].layers.mask;
+      this.snapshot.mask1 = camera.cameras[1].layers.mask;
+      camera.cameras[0].layers.set(CAMERA_LAYER_INSPECT);
+      camera.cameras[1].layers.set(CAMERA_LAYER_INSPECT);
+    }
+
     moveRigSoCameraLooksAtObject(this.rigEl.object3D, this.cameraEl.object3D, this.inspected);
+
+    this.snapshot.audio = getAudio(o);
+    if (this.snapshot.audio) {
+      this.snapshot.audio.applyMatrix(new THREE.Matrix4().identity()); // hack around our matrix optimizations
+      this.snapshot.audio.updateMatrices();
+      this.snapshot.audioTransform.copy(this.snapshot.audio.matrixWorld);
+      scene.audioListener.updateMatrices();
+      setMatrixWorld(this.snapshot.audio, scene.audioListener.matrixWorld);
+    }
   }
 
   uninspect() {
-    this.inspected = null;
+    if (this.inspected) {
+      (getBatch(this.inspected, this.batchManagerSystem) || this.inspected).traverse(disableInspectLayer);
+    }
+    const vrMode = AFRAME.scenes[0].is("vr-mode");
+    const scene = AFRAME.scenes[0];
+    const camera = vrMode ? scene.renderer.vr.getCamera(scene.camera) : scene.camera;
+    camera.layers.mask = this.snapshot.mask;
+    if (vrMode) {
+      camera.cameras[0].layers.mask = this.snapshot.mask0;
+      camera.cameras[1].layers.mask = this.snapshot.mask1;
+    }
+    if (this.inspected) {
+      this.inspected = null;
+    }
     if (this.mode !== CAMERA_MODE_INSPECT) return;
-    this.mode = this.preInspectMode || CAMERA_MODE_FIRST_PERSON;
-    this.preInspectMode = null;
+    this.mode = this.snapshot.mode || CAMERA_MODE_FIRST_PERSON;
+    this.snapshot.mode = null;
+
+    if (this.snapshot.audio) {
+      setMatrixWorld(this.snapshot.audio, this.snapshot.audioTransform);
+    }
   }
+
   tick = (function() {
     const m2 = new THREE.Matrix4();
     const m3 = new THREE.Matrix4();
     const offset = new THREE.Vector3();
-    return function tick() {
+    return function tick(scene) {
+      if (!scene.is("entered")) return;
       this.playerHead = this.playerHead || document.getElementById("avatar-head");
-      if (!AFRAME.scenes[0].is("entered")) return;
```

Comment:
this `tick` method is pretty bulky and does a bunch of work -- we should at least extract the common subexpressions (like the raw rotator) and avoid necessary work. have you tried profiling this?

Reply:
also for the common/perf-sensitive case of 1st person, VR mode, is it possible to avoid some of the matrix copies? it seems like we copy a matrix to rigEl needlessly since we can just read the avatar rig matrix in 1st person mode, and it's a bit unclear to me why we need avatarPOV in VR mode.

Reply:
nevermind re: rig copy, i guess we now need to do that because camera is on that viewing rig

Reply:
i took a quick look and the profile doesn't jump out or anything. but would be nice to eliminate some of this work esp if it's cuttable in VR mode

Reply:
I'll simplify it a bit, but I'm surprised that you find this an expensive `tick`. We can probably get rid of a matrix copy somehow but I imagined this as a tiny blip compared to e.g. copying all the matrices of physics objects to and from the physics world every frame.

The most expensive work is done when first inspecting an object, when we calculate the bounding box, viewing transform, and toggle the camera layers. Besides that is it the number of conditionals that is worrying?